File: lc_soln/0023_merge_k_sorted_linked_lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) {
            return nullptr;
        }
        return split(lists, 0, lists.size() - 1);
    }

    ListNode* split(std::vector<ListNode*>& lists, int l, int r) {
        if (l > r) {
            return nullptr;
        }
        if (l == r) {
            return lists[l];
        }
        int mid = l + (r - l) / 2;
        ListNode* left = split(lists, l, mid);
        ListNode* right = split(lists, mid + 1, r);
        return merge(left, right);
    }

    ListNode* merge(ListNode* left, ListNode* right) {
        ListNode* dummy = new ListNode();
        ListNode* curr = dummy;
        while (left != nullptr && right != nullptr) {
            if (left->val < right->val) {
                curr->next = left;
                left = left->next;
            } else {
                curr->next = right;
                right = right->next;
            }
            curr = curr->next;
        }
        if (left != nullptr) {
            curr->next = left;
        }
        if (right != nullptr) {
            curr->next = right;
        }
        return dummy->next;
    }
};

int main() {
    return 0;
}

File: lc_soln/0023_merge_k_sorted_linked_lists/solution.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution
{
    public ListNode MergeKLists(ListNode[] lists)
    {
        if (lists.Length == 0)
        {
            return null;
        }
        var q = new LinkedList<ListNode>(lists);
        while (q.Count > 1)
        {
            ListNode l1 = q.First.Value;
            q.RemoveFirst();
            ListNode l2 = q.First.Value;
            q.RemoveFirst();
            ListNode l3 = Merge(l1, l2);
            q.AddLast(l3);
        }
        return q.First.Value;
    }

    ListNode Merge(ListNode l1, ListNode l2)
    {
        var dummy = new ListNode();
        ListNode curr = dummy;
        while (l1 is not null && l2 is not null)
        {
            if (l1.val < l2.val)
            {
                curr.next = l1;
                l1 = l1.next;
            }
            else
            {
                curr.next = l2;
                l2 = l2.next;
            }
            curr = curr.next;
        }
        if (l1 is not null)
        {
            curr.next = l1;
        }
        if (l2 is not null)
        {
            curr.next = l2;
        }
        return dummy.next;
    }
}
